content/guides/concepts/hashes.md

```diff
@@ -0,0 +1,59 @@
+---
+title: "Hashes"
+menu:
+    guides:
+        parent: concepts
+---
+
+Hashes are functions that take some arbitrary input and returns a fixed-length value. The particular value depends on the given hash algorithm in use, such as [SHA-1](https://en.wikipedia.org/wiki/SHA-1) (used by Git), [SHA-256](https://en.wikipedia.org/wiki/SHA-2), or [BLAKE2](https://en.wikipedia.org/wiki/BLAKE_(hash_function)#BLAKE2), but a given hash algorithm always returns the same value for a given input. Have a look at the [full list of hash functions](https://en.wikipedia.org/wiki/List_of_hash_functions) for more.
+
+As an example, the input:
+
+```
+Hello world
+```
+
+would be represented by **SHA-1** as:
+
+```
+0x7B502C3A1F48C8609AE212CDFB639DEE39673F5E
+```
+
+However, the exact same input generates the following output using **SHA-256**:
+
+```
+0x64EC88CA00B268E5BA1A35678A1B5316D212F4F366B2477232534A8AECA37F3C
+```
+
+Notice that the second hash is longer than the first one. This is because SHA-1 creates a 160 bit hash, while SHA-256 creates a 256 bit hash. Also, the prepended `0x` is just an indicator that tells us that the following hash is represented as a base 16 (or hexadecimal) number.
+
+Hashes can be represented in different bases (`base2`, `base16`, `base32`, etc.). In fact, IPFS makes use of that as part of its [Content Identifiers](cid.md) and supports mulitiple base representations at the same time, using the [Multibase](https://github.com/multiformats/multibase) protocol.
+
+For example, the SHA-256 hash of "Hello World" from above can be represented as base 32 as:
+
+```
+mtwirsqawjuoloq2gvtyug2tc3jbf5htm2zeo4rsknfiv3fdp46a
+```
+
+An easy way to play around with these in JavaScript is to use the [multibase package](https://github.com/multiformats/js-multibase):
+
+```
+const crypto = require('crypto');
+const multibase = require('multibase');
+
+const hashBytes = crypto.createHash('sha256').update('Hello world').digest();
+multibase.encode('base32', hashBytes).toString();
+```
+
+Just keep in mind that this will output a base 32 string including the multibase prefix.
+
+## Characteristics of cryptographic hashes
+
+Cryptographic hashes come with a couple of very important characteristics:
+
+- **deterministic** - the same input message always returns exactly the same output hash
+- **uncorrelated** - a small change in the message should generate a completely different hash
+- **unique** - it's infeasible to generate the same hash from two different messages
+- **one-way** - it's infeasible to guess or calculate the input message from its hash
+
+Turns out these are a very good fit for a content addressable system like IPFS, where cryptographic hashes are used to identify any kind of content in the network.
```

Comment:
@Mr0grog I tried to address the "why it matters for IPFS" here, I'm not really sure what more to say about it here except for the fact that IPFS uses hashes to identify content and that due to the things given above, hases are a perfect fit for that.

Happy to change that.

Reply:
Hmmmm, I hear you. I think the other concept worth addressing here is why that’s important to *a distributed system* like IPFS. Someone who understands content addressing will probably already know that (or be able to work it out for themselves), but someone reading the basics of hashes here most certainly won’t really know much about content addressing.

So maybe adding something like this?

> It turns out these features also mean we can use a cryptographic hash to *identify* any piece of data: the hash is unique to the data we calculated it from and it’s not too long (a hash is a fixed length, so the SHA-256 hash of a [1 Gigabyte](# "1,000,000,000 bytes") video file is still only [32 bytes](# "That’s 256 bits, hence the name of the hash algorithm!") so sending it around the network doesn’t take up a lot of resources.
> 
> That’s critical for a distributed system like IPFS, where we want to be able to store and retrieve data from many places. A computer running IPFS can ask all the computers near it whether they have a file with a particular hash and, if one of them does, they send back the whole file. Without a short, unique value like a cryptographic hash, that wouldn’t be possible. This technique is called “content addressing” — because the content itself is used to form an address, rather than information about the computer and disk location it’s stored at.